### PR TITLE
Simple date format has an issue with Android M, N and O so using basic string operations

### PR DESCRIPTION
**Pull request by VinayGuthal**
Fixes https://github.com/firebase/firebase-android-sdk/issues/2622 as per https://b.corp.google.com/issues/110848122#comment17

### CODE REVIEW

File: firebase-common/src/main/java/com/google/firebase/heartbeatinfo/HeartBeatInfoStorage.java [hunk 1]
```diff
@@ -127,10 +127,18 @@ synchronized void clearStoredHeartBeats() {
     sharedPreferences.edit().remove(HEART_BEAT_COUNT_TAG).apply();
   }
 
-  static boolean isSameDateUtc(long base, long target) {
+  synchronized static boolean isSameDateUtc(long base, long target) {
     Date baseDate = new Date(base);
     Date targetDate = new Date(target);
-    return !(FORMATTER.format(baseDate).equals(FORMATTER.format(targetDate)));
+    String baseDateString = baseDate
+            .toInstant()
```
Comment (vkryachko) [inline firebase-common/src/main/java/com/google/firebase/heartbeatinfo/HeartBeatInfoStorage.java:134]: `toInstant()` is not available in Java < 8 so can only be used on devices with API < 24/

**Review by vkryachko — CHANGES_REQUESTED** [review]

File: firebase-common/src/main/java/com/google/firebase/heartbeatinfo/DefaultHeartBeatInfo.java [hunk 2]
```diff
@@ -93,7 +93,7 @@ private DefaultHeartBeatInfo(Context context, Set<HeartBeatConsumer> consumers)
           HeartBeat heartBeat;
           for (SdkHeartBeatResult sdkHeartBeatResult : sdkHeartBeatResults) {
             shouldSendGlobalHeartBeat =
-                HeartBeatInfoStorage.isSameDateUtc(
+                !HeartBeatInfoStorage.isSameDateUtc(
```
[the remark this answers is not in the document]
Comment (VinayGuthal) [reply firebase-common/src/main/java/com/google/firebase/heartbeatinfo/DefaultHeartBeatInfo.java:96]: actually isSameDateUtc was returning the reverse value of what the name suggests.. So fixed that

File: firebase-common/src/main/java/com/google/firebase/heartbeatinfo/HeartBeatInfoStorage.java [hunk 3]
```diff
@@ -127,10 +127,18 @@ synchronized void clearStoredHeartBeats() {
     sharedPreferences.edit().remove(HEART_BEAT_COUNT_TAG).apply();
   }
 
-  static boolean isSameDateUtc(long base, long target) {
+  synchronized static boolean isSameDateUtc(long base, long target) {
```
[the remark this answers is not in the document]
Comment (VinayGuthal) [reply firebase-common/src/main/java/com/google/firebase/heartbeatinfo/HeartBeatInfoStorage.java:130]: only one thread needs to be able to get any sort of timezone information for this to work

[the remark this answers is not in the document]
Comment (vkryachko) [reply firebase-common/src/main/java/com/google/firebase/heartbeatinfo/HeartBeatInfoStorage.java:130]: what do you mean? looking at it, this is a completely pure function that takes 2 arguments and returns a boolean and does not use any shared mutable state outside of its body, what relies on it being synchronized?

File: firebase-common/src/main/java/com/google/firebase/heartbeatinfo/DefaultHeartBeatInfo.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (vkryachko) [reply firebase-common/src/main/java/com/google/firebase/heartbeatinfo/DefaultHeartBeatInfo.java:96]: While it makes sense, it sounds unrelated to the bug in question, can we address this issue in a separate PR?

Also, might make sense to rename the method and not have a negation for less cognitive load on the reader.

**Comment by vkryachko** [comment]
Also, let's reference the underlying system bug in the code and in the PR message:
https://issuetracker.google.com/issues/110848122#comment17

File: firebase-common/src/main/java/com/google/firebase/heartbeatinfo/DefaultHeartBeatInfo.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (VinayGuthal) [reply firebase-common/src/main/java/com/google/firebase/heartbeatinfo/DefaultHeartBeatInfo.java]: Okay removed the change and created https://b.corp.google.com/issues/192872326 to track this

File: firebase-common/src/main/java/com/google/firebase/heartbeatinfo/HeartBeatInfoStorage.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (VinayGuthal) [reply firebase-common/src/main/java/com/google/firebase/heartbeatinfo/HeartBeatInfoStorage.java]: As per https://b.corp.google.com/issues/110848122#comment17 the mutable state is in the enum switch.

File: firebase-common/src/main/java/com/google/firebase/heartbeatinfo/HeartBeatInfoStorage.java (hunk 1 above)
Comment (VinayGuthal) [reply firebase-common/src/main/java/com/google/firebase/heartbeatinfo/HeartBeatInfoStorage.java]: Removed it

**Pull request closed without merging**